### PR TITLE
Altitude hold controller

### PR DESCRIPTION
**Pull request by alphaville**
About:
- complete implementation of altitude hold controller
- tested on quadcopter
- credit to @gunturiCM

## Tests

Hardware tests done

### CODE REVIEW

File: src/radio.cpp [hunk 1]
```diff
@@ -34,9 +34,11 @@
 
 namespace bzzz
 {
-    Radio::Radio(){};
+    Radio::Radio(bool replyWithIMUData /*=flase*/){
+        this->m_replyWithIMUData = replyWithIMUData;
+    };
     
-    bool Radio::readPiData(void)
+    bool Radio::readPiData(float q1, float q2, float q3, float ax, float ay, float az, float motorFL, float motorFR, float motorBL, float motorBR)
```
Comment (alphaville) [inline src/radio.cpp:41]: I don't understand this method. Previously, it used to simply and literally read the data coming from Pi. What are these input arguments? Also I see that this method **sends** instead of reading.

File: include/radio.hpp [hunk 2]
```diff
@@ -64,15 +64,26 @@ namespace bzzz
...
 
+        /**
+         * Boolean to enable sending IMU data to Pi whenever ESP receives radio data
+         * if True: replies with IMU data;  
+        */
+       bool m_replyWithIMUData;
+
     public:
-        Radio();
+        Radio(bool replyWithIMUData = false);
```
Comment (alphaville) [inline include/radio.hpp:74]: This is undocumented. Moreover, I don't understand how this works.

File: src/radio.cpp [hunk 3]
```diff
@@ -79,11 +81,41 @@ namespace bzzz
...
+        Serial.print(az);
+        Serial.print(' ');
+        Serial.print(motorFL);
+        Serial.print(' ');
+        Serial.print(motorFR);
+        Serial.print(' ');
+        Serial.print(motorBL);
+        Serial.print(' ');
+        Serial.println(motorBR);
+        // String flightData = "FD: " + String(q1) + " " + String(q2) + " " + String(q3)
```
Comment (alphaville) [inline src/radio.cpp:114]: remove commented-out code

File: include/ahrs.hpp [hunk 4]
```diff
@@ -133,6 +133,13 @@ namespace bzzz
         */
         float currentYawRad();
 
+        /**
+         * Get the acclerometer values
```
Comment (alphaville) [inline include/ahrs.hpp:137]: acclerometer -> accelerometer

File: src/main.cpp [hunk 5]
```diff
@@ -62,11 +64,11 @@ void setup()
 void setGainsFromRcTrimmers()
 {
   controller.setQuaternionGain(
-      -radio.trimmerVRAPercentage() * RADIO_TRIMMER_MAX_QUATERNION_XY_GAIN);
+      - 0.470 * RADIO_TRIMMER_MAX_QUATERNION_XY_GAIN);
```
Comment (alphaville) [inline src/main.cpp:67]: These values are now fixed. Why don't you define them as new constants. Constants like `RADIO_TRIMMER_MAX_QUATERNION_XY_GAIN` are not needed any more.

File: src/main.cpp [hunk 6]
```diff
@@ -125,18 +127,34 @@ void loop()
   Quaternion relativeQuaternion = currentQuaternion - initialQuaternion;
   Quaternion attitudeError = referenceQuaternion - relativeQuaternion; // e = set point - measured
 
+  IMUData[0] = relativeQuaternion[1];
+  IMUData[1] = relativeQuaternion[2];
+  IMUData[2] = relativeQuaternion[3];
+  ahrs.getAcclerometerValues(IMUData + 3);
+
+  // Serial.print("FD: ");
```
Comment (alphaville) [inline src/main.cpp:135]: remove all commented-out code

File: src/radio.cpp [hunk 7]
```diff
@@ -34,9 +34,11 @@
 
 namespace bzzz
 {
-    Radio::Radio(){};
+    Radio::Radio(bool replyWithIMUData /*=flase*/){
```
Comment (alphaville) [inline src/radio.cpp:37]: flase -> false

File: include/radio.hpp [hunk 8]
```diff
@@ -145,6 +156,17 @@ namespace bzzz
...
+        /**
+         * @brief send IMU data quaternions and acclerometer values to Pi
+         * @param q1
+         * @param q2
+         * @param q3
+         * @param ax
+         * @param ay
+         * @param az
+        */
+        void sendIMUDataToPi(float q1, float q2, float q3, float ax, float ay, float az, float motorFL, float motorFR, float motorBL, float motorBR);
```
Comment (alphaville) [inline include/radio.hpp:168]: This method is called `sendIMUDataToPi` but it also sends the motor commands? Are the motor commands needed to Pi?

File: raspberry/bzzz/controllers/altitude_LQR.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
+# Imports
+import numpy as np
+import control as ctrl
+from math import cos
+
+class LQR:
+    def __init__(self, sampling_frequency = 10, initial_alpha_t = 0, initial_beta_t = 0) -> None:
```
Comment (alphaville) [inline raspberry/bzzz/controllers/altitude_LQR.py:7]: constructors do not return `None`

File: raspberry/bzzz/controllers/altitude_LQR.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+            [self.__Ts*self.__beta_t]
+        ])
+
+        # reference tracker parameters
+        self.__x_and_u_bar = np.zeros((3, 1))  # equilibrium points "The hover estimate"
+        self.__reference_tracker_dynamics = np.zeros((self.__A.shape[0] + self.__C.shape[0], self.__A.shape[1] + self.__B_t.shape[1]), dtype=np.float64)
+        self.__reference_tracker_matrix = np.zeros((3, 1))
+
+        # LQR matrices
+        self.__Q = np.array([
```
Comment (alphaville) [inline raspberry/bzzz/controllers/altitude_LQR.py:37]: np.eye(2)

File: raspberry/bzzz/controllers/altitude_LQR.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    def __calculate_stabilising_gain(self):
+        """Calculate the stabilising feedback gain using LQR method, call this function whenever the system dynamics are updated 
+        """
+        _, _, kappa = ctrl.dare(self.__A, self.__B_t, self.__Q, self.__R)
+        self.__kappa[:, :] = -np.copy(kappa)
+
+    def control_action(self, current_states_z_and_vz: np.array, alpha_t = 0, beta_t = 0, reference_altitude_mts = 1, recalculate_dynamics = False, pitch_rad=0, roll_rad=0, k11=None, k12=None):
+        """Calculate and return the normalized control action
+
+        :param current_states_z_and_vz: current states of altitude in mts and velocity in m/s in a 2x1 numpy array
```
Comment (alphaville) [inline raspberry/bzzz/controllers/altitude_LQR.py:125]: meters or m (not mts)

File: raspberry/main.py [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+    # objects declaration
+    kf = KalmanFilter(sampling_frequency=sampling_frequency,
+                      initial_Tt=0,
+                      x_tilde_0=np.array([[0], [0], [10], [-9.81]]),
+                      P_0=np.diagflat([1, 1, 1, 0.01]),
+                      cache_values=True)
+    lqr = LQR(sampling_frequency=sampling_frequency,
+              initial_alpha_t=10,
+              initial_beta_t=-9.81)
+    tof = TimeOfFlightSensor(use_sleep=-1,
```
Comment (alphaville) [inline raspberry/main.py:40]: use_sleep can only be `True` or `False` (not -1)

File: raspberry/bzzz/sensors/time_of_flight_sensor.py [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+    def __init__(self,
+                 num_latest_readings_to_keep: int = 3,
+                 use_sleep=False,
+                 cache_altitude=False,
+                 use_outlier_detection=False,
+                 abs_outlier_diff_thres=500):
+        """Time of flight sensor class initialization.
+
+        :param num_latest_readings_to_keep: Window length for median filter, defaults to 3
+        :param use_sleep: Indicator to use sleep and wait for the next measurement; Not recommended as this will halt the system, defaults to False
```
Comment (alphaville) [inline raspberry/bzzz/sensors/time_of_flight_sensor.py:19]: Since `use_sleep` is not recommended and should not be used, why have it in the code?

Comment (alphaville) [reply raspberry/bzzz/sensors/time_of_flight_sensor.py:19]: To be clear, I'm saying that we shouldn't have it in the code

File: raspberry/bzzz/estimators/altitude_Kalman_filter.py [hunk 14]
```diff
@@ -1,9 +1,16 @@
...
+
 
 class KalmanFilter:
-    """This is a Kalman filter for estimating altitude hold related parameters.
-    """
-    def __init__(self, sampling_frequency=10, initial_Tt=0., x_tilde_0=np.zeros((4, 1)), P_0=np.eye(4, 4)*100, cache_values=False):
+    def __init__(self,
+                 sampling_frequency=10,
+                 initial_Tt=0.,
+                 x_tilde_0=np.zeros((4, 1)),
```
Comment (alphaville) [inline raspberry/bzzz/estimators/altitude_Kalman_filter.py:9]: Document what the state is

Comment (gunturiCM) [reply raspberry/bzzz/estimators/altitude_Kalman_filter.py:15]: Done!

File: raspberry/bzzz/sensors/time_of_flight_sensor.py (hunk 13 above)
Comment (gunturiCM) [reply raspberry/bzzz/sensors/time_of_flight_sensor.py]: Removed!

File: raspberry/main.py (hunk 12 above)
Comment (gunturiCM) [reply raspberry/main.py]: Done!

File: raspberry/bzzz/controllers/altitude_LQR.py (hunk 11 above)
Comment (gunturiCM) [reply raspberry/bzzz/controllers/altitude_LQR.py]: Done!

File: raspberry/bzzz/controllers/altitude_LQR.py (hunk 10 above)
Comment (gunturiCM) [reply raspberry/bzzz/controllers/altitude_LQR.py]: Done!

File: raspberry/bzzz/controllers/altitude_LQR.py (hunk 9 above)
Comment (gunturiCM) [reply raspberry/bzzz/controllers/altitude_LQR.py]: Removed.

File: include/radio.hpp (hunk 8 above)
Comment (gunturiCM) [reply include/radio.hpp]: Renamed the function `sendIMUDataToPi` to `sendFlightDataToPi`. Currently, we need the motor data for plotting purposes.

File: src/main.cpp (hunk 6 above)
Comment (gunturiCM) [reply src/main.cpp]: Done!

File: src/main.cpp (hunk 5 above)
Comment (gunturiCM) [reply src/main.cpp]: Done! Now the constant gains are defined as `QUATERNION_XY_GAIN`, `OMEGA_XY_GAIN`, and `OMEGA_Z_GAIN`.

File: src/radio.cpp (hunk 3 above)
Comment (gunturiCM) [reply src/radio.cpp]: Done!

File: include/radio.hpp (hunk 2 above)
Comment (gunturiCM) [reply include/radio.hpp]: Renamed `replyWithIMUData` to `replyWithFlightData`. Documentation added.

File: src/radio.cpp (hunk 1 above)
Comment (gunturiCM) [reply src/radio.cpp:41]: We pass the current drone's attitude `Quaternions`, linear accelerations, and motor `PWM` signals to `readPiData` function which invokes `sendFlightDataToPi` (previously declared as `sendIMUDataToPi`) if radio data from `Pi` is received.

File: src/radio.cpp (hunk 7 above)
Comment (gunturiCM) [reply src/radio.cpp]: Professor, I didn't get you.

File: include/ahrs.hpp (hunk 4 above)
Comment (gunturiCM) [reply include/ahrs.hpp:137]: I didn't get this comment either.

Comment (gunturiCM) [reply include/ahrs.hpp]: Done!

File: src/radio.cpp (hunk 7 above)
Comment (gunturiCM) [reply src/radio.cpp]: Done!

File: src/radio.cpp (hunk 1 above)
Comment (alphaville) [reply src/radio.cpp:41]: It is unintuitive - the method is called "read" but it "sends". What does it read?

File: raspberry/bzzz/__init__.py [hunk 15]
```diff
@@ -0,0 +1,5 @@
+# import bzzz.controllers
+import bzzz.read_sbus
+import bzzz.sensors
+# import bzzz.thread_this
```
Comment (alphaville) [inline raspberry/bzzz/__init__.py:4]: remove all commented-out code

File: raspberry/bzzz/controllers/altitude_LQR.py [hunk 16]
```diff
@@ -0,0 +1,147 @@
+# Imports
+import numpy as np
+import control as ctrl
+from math import cos
+
+class LQR:
+    def __init__(self, sampling_frequency = 10, initial_alpha_t = 0, initial_beta_t = 0):
```
Comment (alphaville) [inline raspberry/bzzz/controllers/altitude_LQR.py:7]: write documentation

File: raspberry/bzzz/estimators/altitude_Kalman_filter.py [hunk 17]
```diff
@@ -22,26 +36,26 @@ def __init__(self, sampling_frequency=10, initial_Tt=0., x_tilde_0=np.zeros((4,
...
+        self.__sigma_MU = overwrite_sigma_MU
 
         # Time update
         self.__x_TU = x_tilde_0
         self.__sigma_TU = P_0
 
-        # Point to a function that does nothing if caching is not enabled.
-        self.__cache_MU_values = lambda : ()
-        self.__cache_TU_values = lambda : ()
+        self.__cache_MU_values = lambda: ()
```
Comment (alphaville) [inline raspberry/bzzz/estimators/altitude_Kalman_filter.py:55]: I still find this overly complicated. These are caches and caches are arrays (or lists if you prefer) - not lambdas.

File: raspberry/bzzz/estimators/altitude_Kalman_filter.py [hunk 18]
```diff
@@ -71,64 +85,90 @@ def __update_At(self):
...
 
         :param Tt: Current normalized throttle reference.
+        :param pitch_rad: current drone pitch in radians.
+        :param roll_rad: current drone roll in radians.
         """
-        self.__Tt = Tt
+        self.__Tt = math.cos(pitch_rad) * math.cos(roll_rad) * float(Tt)
 
-    def measurement_update_cache(self):
+    def MU_cache(self):
```
Comment (alphaville) [inline raspberry/bzzz/estimators/altitude_Kalman_filter.py:106]: I thought you had renamed this to `measurement_update_cache`. Have you pushed the changes?

File: raspberry/bzzz/scheduler.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    _function_ID = "function_ID"
+    _function_obj = "function_obj"
+    _function_call_frequency = "function_call_frequency"
+    _function_call_count = "function_call_count"
+    _function_thread = "function_thread"
+    _num_times_called = "num_times_called"
+    _time_at_last_call_ns = "time_at_last_call_ns"
+    _dummy_func = "dummy_func"
+
+    def __init__(self, use_threading=False) -> None:
```
Comment (alphaville) [inline raspberry/bzzz/scheduler.py:16]: Constructors do not return `None`. Make sure you correct this throughout the entire code.

File: raspberry/bzzz/scheduler.py
Comment (alphaville) [inline raspberry/bzzz/scheduler.py:1]: The class `Scheduler` is not adequately documented

File: raspberry/bzzz/sensors/pressure_sensor.py
Comment (alphaville) [inline raspberry/bzzz/sensors/pressure_sensor.py:1]: Has this been tested?

File: raspberry/bzzz/sensors/time_of_flight_sensor.py [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+from time import time_ns, sleep
+from datetime import datetime
+import VL53L0X
+import os
+
+
+class TimeOfFlightSensor:
+    # NOTE: This class fetches altitude in millimeters and this class requires i2c interfacing with the sensor.
+    def __init__(self,
+                 num_latest_readings_to_keep: int = 3,
```
Comment (alphaville) [inline raspberry/bzzz/sensors/time_of_flight_sensor.py:11]: wouldn't `median_filter_length` be a better name?

File: raspberry/bzzz/sensors/time_of_flight_sensor.py [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+                 update_measurement_at_fixed_rate=False,
```
Comment (alphaville) [inline raspberry/bzzz/sensors/time_of_flight_sensor.py:12]: Don't you want it to give you measurements at a constant rate? What is the current update rate? Can it be configured?

File: raspberry/bzzz/sensors/time_of_flight_sensor.py [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+        """
+        status = self.__update_altitude()
+        if status == -1:
+            return -1
+        return self._current_altitude
+
+    @altitude.setter
+    def altitude(self, val):
+        """Setter for altitude measurement.
+        NOTE: only use this for testing purposes, do not manually
```
Comment (alphaville) [inline raspberry/bzzz/sensors/time_of_flight_sensor.py:102]: should this method be removed?

File: raspberry/data/testdata.csv
Comment (alphaville) [inline raspberry/data/testdata.csv:1]: add a header line to the CSV file

File: src/radio.cpp (hunk 1 above)
Comment (gunturiCM) [reply src/radio.cpp:41]: 1. It reads Radio reference data from Pi.
2. It sends flight data (current drone attitude Quaternion vector part, current IMU accelerometer values, current motor PWM values) to R-Pi. 
3. This `readPiData` calls `sendFlightDataToPi` to send flight data. 
I will make some changes to avoid this confusion.

File: raspberry/main.py [hunk 23]
```diff
@@ -0,0 +1,371 @@
...
+        Tref_t[0] = (rc.throttle_reference_percentage() - 1000)/900
+        cache_values()  # call to cache values
+
+    def process_ESP_data():
+        """Process ESP data, this function does two jobs:
+        1. Receives flight data as a string of space seperated values formatted as "FD: q1 q2 q3 ax ay az".
+        2. Checks the received flight data for corruption. 
+           If string is not None and starts with "FD:", and if there are 7 space seperated values
+           then convert the 7 values to floats and update shared variables.
+        NOTE: might want to include a check for value data corruption before converting to floats,
```
Comment (alphaville) [inline raspberry/main.py:205]: Why don't you make an issue about this? Otherwise, we'll forget.

File: raspberry/main.py [hunk 24]
```diff
@@ -0,0 +1,371 @@
...
+            print(
+                f"alphan hat: {alpha_hat[0]} beta hat: {beta_hat[0]} k11: {-LQR_Q11_gain[0]} k12: {-LQR_Q22_gain[0]} Tref_LQR: {Tref_t[0]} alt_hat: {z_hat[0]} alt_ref: {altitude_ref_mts[0]}")
+
+    if enable_caching[0]:
+        time_before_thread_starts[0] = time_ns()
+
+    # schedule the necessary functions
+    scheduler.schedule("process_radio_data",
+                       process_radio_data,
+                       function_call_frequency=50,
```
Comment (alphaville) [inline raspberry/main.py:330]: should this be `function_call_frequency=sampling_frequency`? (instead of hard-coding the value 50)

File: raspberry/bzzz/scheduler.py [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+        # This variable is a dict with the following structure
```
Comment (alphaville) [inline raspberry/bzzz/scheduler.py:17]: Is it a dictionary or a list of dictionaries?

File: raspberry/bzzz/scheduler.py [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+        # {function_name: {function_ID: int, function_obj: obj, function_call_frequency: Hz, function_call_count: int, function_thread: threading.Timer or None, num_times_called: int, time_at_last_call_ns: int}}
+        self.__functions_to_run = {}
+        self.__last_function_ID = -1
+        self.__use_threading = use_threading
+        self.__scheduled_events = []
+
+        self.__functions_to_run[Scheduler._dummy_func] = {
```
Comment (alphaville) [inline raspberry/bzzz/scheduler.py:24]: use one line per key-value pair for the sake of readability

File: raspberry/bzzz/scheduler.py [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+            Scheduler._function_ID: self.__last_function_ID, Scheduler._function_obj: -1,
+            Scheduler._function_call_frequency: -1, Scheduler._function_call_count: -1,
+            Scheduler._function_thread: -1, Scheduler._num_times_called: -1, Scheduler._time_at_last_call_ns: -1}
+
+    def schedule(self, function_name: str, function_obj, function_call_frequency, function_call_count=0):
```
Comment (alphaville) [inline raspberry/bzzz/scheduler.py:29]: Regarding `function_name: str`, I see that you have documented the datatype this way in a very few cases. Since this is not something we do uniformly throughout the code, I'd suggest that we abandon it and mention the data type in the docstrings.

File: raspberry/bzzz/simulations/altitude_dynamics.py
Comment (alphaville) [inline raspberry/bzzz/simulations/altitude_dynamics.py:1]: I'm a little confused about this file. What does it do? If it's WIP, it shouldn't be included in the PR.

Comment (alphaville) [inline raspberry/bzzz/simulations/altitude_dynamics.py:1]: I also think that `simulations` should not be inside `raspberry`. Wouldn't it make more sense to keep the code for simulations separately? It now looks as if we're running simulations on the pi.

File: raspberry/bzzz/estimators/altitude_Kalman_filter.py [hunk 28]
```diff
@@ -1,9 +1,22 @@
...
+    Where, alpha and beta are parameters that relate throttle reference percentage with vertical accleration as follows,
+    ===> a_z = alpha * Throttle_reference_percentage_along_z + beta
+    and Throttle_reference_percentage_along_z = Throttle_reference_percentage * cos (current_drone_pitch_rad) * cos (current_drone_roll_rad)
     """
-    def __init__(self, sampling_frequency=10, initial_Tt=0., x_tilde_0=np.zeros((4, 1)), P_0=np.eye(4, 4)*100, cache_values=False):
+    def __init__(self,
+                 sampling_frequency=10,
+                 initial_Tt=0.,
+                 x_tilde_0=np.zeros((4, 1)),
+                 P_0=np.eye(4, 4)*100,
```
Comment (alphaville) [inline raspberry/bzzz/estimators/altitude_Kalman_filter.py:16]: Isn't it `np.eye(4)`?

File: raspberry/bzzz/estimators/altitude_Kalman_filter.py (hunk 17 above)
Comment (gunturiCM) [reply raspberry/bzzz/estimators/altitude_Kalman_filter.py:55]: These are functions, which when called cache values into lists. If we configure to cache values, they will point to a function that does cache, otherwise, they point to this empty function. I will change these to be less complicated.

File: raspberry/bzzz/sensors/pressure_sensor.py
Comment (gunturiCM) [reply raspberry/bzzz/sensors/pressure_sensor.py:1]: Yes, this has been tested.

File: raspberry/bzzz/sensors/time_of_flight_sensor.py (hunk 20 above)
Comment (gunturiCM) [reply raspberry/bzzz/sensors/time_of_flight_sensor.py:11]: Done!

File: raspberry/bzzz/sensors/time_of_flight_sensor.py (hunk 22 above)
Comment (gunturiCM) [reply raspberry/bzzz/sensors/time_of_flight_sensor.py:102]: Yes, it has no use now. I will remove it.

File: raspberry/main.py (hunk 23 above)
Comment (gunturiCM) [reply raspberry/main.py:205]: Done! Issue #108 opened to address this.

File: raspberry/bzzz/sensors/time_of_flight_sensor.py (hunk 21 above)
Comment (gunturiCM) [reply raspberry/bzzz/sensors/time_of_flight_sensor.py:12]: Currently, we are sampling at `50 Hz` and a `scheduler` is taking care of this constant update rate.  A **custom** update rate cannot be configured by using `update_measurement_at_fixed_rate` because this will only ensure that the sensor is running at the maximum update frequency possible for the current `ranging configuration`. For example, say that the maximum sampling rate is `20 Hz` in `long range mode`, using `update_measurement_at_fixed_rate` only ensures that the sensor is sampled at `20 Hz` (could be less) whilst disabling this and using `scheduler` enables us to sample at **any rate lower than `20 Hz`**.

File: raspberry/main.py (hunk 24 above)
Comment (gunturiCM) [reply raspberry/main.py:330]: No, we might want to run the other process at different rates than `LQR and KF`. Although for now, they happen to have the same `function_call_frequency`, they should not be set to `sampling_frequency`. Because in future we will have to change `sampling_frequency` to a different value based on what the new **better quality** sensors support, which could be something different than `50 Hz`. But we will always want to run communication tasks between `ESP` and `R-Pi` at `50 Hz` and hence hard-coded values.

File: raspberry/bzzz/scheduler.py (hunk 25 above)
Comment (gunturiCM) [reply raspberry/bzzz/scheduler.py:17]: It is a dictionary of dictionaries.

File: src/radio.cpp (hunk 1 above)
Comment (gunturiCM) [reply src/radio.cpp]: Done!

File: raspberry/bzzz/__init__.py (hunk 15 above)
Comment (gunturiCM) [reply raspberry/bzzz/__init__.py]: Done!

**Comment by alphaville** [comment]
@jamie-54 since there are lots of changes in this PR, can you review it yourself before we merge it?

File: raspberry/bzzz/estimators/altitude_Kalman_filter.py (hunk 17 above)
Comment (gunturiCM) [reply raspberry/bzzz/estimators/altitude_Kalman_filter.py]: Done!

**Comment by alphaville** [comment]
A general comment: I'm not sure whether we want to address this now or later, but the class `RC` does not only read the values from the RC, but also sends data from the Pi to ESP32, which is none of its business. If we can't fix this now, let's create an issue - @gunturiCM

File: raspberry/main.py [hunk 29]
```diff
@@ -0,0 +1,416 @@
...
+            throttle_ref_from_LQR[0] = max(
+                1000, min(throttle_ref_from_LQR[0]*900, 600) + 1000)
+            Tref_t[0] = (throttle_ref_from_LQR[0] - 1000)/900
+            print_debug(
+                f"alphan hat: {alpha_hat[0]} beta hat: {beta_hat[0]} k11: {-gain_kp_from_rc[0]} k12: {-gain_kd_from_rc[0]} Tref_LQR: {Tref_t[0]} alt_hat: {z_hat[0]} alt_ref: {altitude_ref_mts[0]}")
+
+    if enable_caching[0]:
+        time_before_thread_starts[0] = time_ns()
+
+    # initialisation checks: are the switches at the correct positions before flying?
```
Comment (alphaville) [inline raspberry/main.py:346]: On the one hand, it looks like these initialisation checks work: the code gets stuck there until the switches are at the correct positions. However, it seems that we can un-kill and start to fly. I'm not sure why.

**Comment by alphaville** [comment]
> A general comment: I'm not sure whether we want to address this now or later, but the class `RC` does not only read the values from the RC, but also sends data from the Pi to ESP32, which is none of its business. If we can't fix this now, let's create an issue - @gunturiCM

To avoid keeping this open for too long, let's take care of the final details (documentation and some final checks). I'll open an issue about this

**Comment by gunturiCM** [comment]
> > A general comment: I'm not sure whether we want to address this now or later, but the class `RC` does not only read the values from the RC, but also sends data from the Pi to ESP32, which is none of its business. If we can't fix this now, let's create an issue - @gunturiCM
> 
> To avoid keeping this open for too long, let's take care of the final details (documentation and some final checks). I'll open an issue about this

The opened issue is #109, this and #110  are expected to be addressed in #111.

File: raspberry/main.py [hunk 30]
```diff
@@ -0,0 +1,405 @@
...
+    acc = [0., 0., 0.]
+    euler = [0., 0., 0.]
+    motor_PWM = [0., 0., 0., 0.]  # FL, FR, BL, BR
+    channel_data = [""]
+    KF_data = [0., 0., 0., 0.]
+
+    switch_a_status = [True]
+    is_data_log_kill = [False]
+    is_kill = [False]
+    allow_data_logging = [True]
```
Comment (alphaville) [inline raspberry/main.py:75]: Add a comment here or we'll soon forget what `allow_data_logging` is for

File: raspberry/main.py [hunk 31]
```diff
@@ -0,0 +1,405 @@
...
+        1. Receives flight data as a string of space seperated values formatted as "FD: q1 q2 q3 ax ay az".
+        2. Checks the received flight data for corruption. 
+           If string is not None and starts with "FD:", and if there are 7 space seperated values
+           then convert the 7 values to floats and update shared variables.
+        """
+        # process ESP data
+        flight_data_string = rc.receive_data_from_ESP()
+        if flight_data_string is not None and "FD:" in flight_data_string:
+            flight_data = flight_data_string.strip().split()
+            if len(flight_data) == 11:
```
Comment (alphaville) [inline raspberry/main.py:214]: @gunturiCM This hasn't been tested yet. We need to flash the ESP and do another test. I hope we can merge this today.

File: raspberry/bzzz/read_sbus/read_sbus_from_GPIO_receiver.py [hunk 32]
```diff
@@ -1,125 +1,139 @@
...
+        while self.ser.inWaiting() > 0:
+            try:
+                line = self.ser.readline().decode('ascii').rstrip()
+                return line
+            except UnicodeDecodeError as e:
+                print(f"UnicodeDecodeError {e}, retrying....")
+        else:
+            return None
+
+    def get_radio_data_parse_and_send_to_ESP(self, return_channel_date=False, force_send_fake_data=False, fake_data="", over_write_throttle_ref_to=-1):
```
Comment (alphaville) [inline raspberry/bzzz/read_sbus/read_sbus_from_GPIO_receiver.py]: `return_channel_date` should be `return_channel_data`?

File: raspberry/main.py (hunk 30 above)
Comment (gunturiCM) [reply raspberry/main.py:86]: Done!

File: raspberry/bzzz/read_sbus/read_sbus_from_GPIO_receiver.py (hunk 32 above)
Comment (gunturiCM) [reply raspberry/bzzz/read_sbus/read_sbus_from_GPIO_receiver.py]: Done!

**Review by gunturiCM — APPROVED** [review]
Everything looks good.

File: raspberry/main.py (hunk 31 above)
Comment (gunturiCM) [reply raspberry/main.py:225]: This has been tested and the logging works.

**Comment by alphaville** [comment]
Let's merge this now and we'll come back to it for a mild refactoring later. If anything comes up, let us create an issue.
